src/t_set.c

```diff
@@ -490,8 +505,8 @@ void setTypeConvert(robj *setobj, int enc) {
         setobj->encoding = OBJ_ENCODING_HT;
         setobj->ptr = d;
     } else if (enc == OBJ_ENCODING_LISTPACK) {
-        /* Preallocate the minimum one byte per element */
-        size_t estcap = setTypeSize(setobj);
+        /* Preallocate the minimum two bytes per element */
+        size_t estcap = cap * 2;
```

Comment:
What about using `intsetBlobLen()`?

Reply:
not sure it's a good indication either.
e.g. an intset with 120 values lower than 130, and then one at around 1,000,000, will allocate a lot of memory in intset, but will be a lot smaller in listpack.

Reply:
Yeah, you are right, at this point the listpack(LP_ENCODING_7BIT_UINT_ENTRY_SIZE(2)*120 + LP_ENCODING_32BIT_INT_ENTRY_SIZE(6)) is only half the size of the intset(sizeof(int32_t) * 121).
But I'm not sure if that's what we want, we should be able to allocate enough space and then shrink it, which should be faster than multiple `realloc()`.

Reply:
well, you're right. we better pre-allocate something a bit too big.
i guess since we can't know the exact allocation size, we better aim too high and then shrink, as long as we don't aim too high and run out of memory.
so intsetBlobLen could be a good idea (when accompanied with an appropriate comment)

Reply:
I think it's weird to use intset as an estimate since is implemented in a totally different way. I'd rather calculate an upper bound in a way similar to the safe-to-add check we do in setTypeAddAux:

```C
            size_t maxelelen = intsetLen(set->ptr) == 0 ?
                0 : max(sdigits10(intsetMax(set->ptr)),
                        sdigits10(intsetMin(set->ptr)));
            if (intsetLen((const intset*)set->ptr) < server.set_max_listpack_entries &&
                len <= server.set_max_listpack_value &&
                maxelelen <= server.set_max_listpack_value &&
                lpSafeToAdd(NULL, maxelelen * intsetLen(set->ptr) + len))
            {
                /* In the "safe to add" check above we assumed all elements in
                 * the intset are of size maxelelen. This is an upper bound. */
```

... but we can do better if we add a function to listpack to compute the size of a listpack with N repetitions of a value M. Then we can do something like

```C
        estcap = max(lpEstimateBytes(cap, intsetMin(setobj->ptr)),
                     lpEstimateBytes(cap, intsetMax(setobj->ptr));
```

Reply:
@sundb interesting. But in an intset, all elements have the same encoding. In a listpack, each element can have different encodings.

This means that intsetBlobLen is not more accurate than using intsetMin and intsetMax times the number of elements to estimate the size of a listpack.